### ISSUE TITLE
Captcha challenge breaking the login process

### ISSUE THREAD

**Issue by Jesus-M** [badger707/esb-smart-meter-reading-automation#8]
Hi!

And apparently ESB doesn't like scripts downloading the files (but they still don't provide an official way to do it), so they introduced a Captcha challenge that, I think, is breaking the login process.

**Comment by badger707** [comment]
Yeah, captcha, another reason why we cant have nice things when we really want/need them in this modern country. I'm experimenting with Coral TPU and some open source captcha solvers at the moment, stay tuned. I'm sure ESB is aware of this repo and it took them over 1year to realise whats going on & close the "loophole".

**Comment by Jesus-M** [comment]
Apparently they removed the captcha challenge now, but probably made some changes in the site, because the code will not work.

**Comment by vitopiepoli** [comment]
I tried with selenium chrome driver to access the source two days ago and
it had the captcha activated.

**Comment by badger707** [comment]
I think I found the cure to this. It looks like they have twisted the process and all request must be made in specific form/shape/headers & sprinkled with specific cookies & tokens per each request (3 requests). I'm half way there in mimicking "true web browser", will try to finish over the weekend and will share updates/status.

**Comment by dermotw** [comment]
> Yeah, captcha, another reason why we cant have nice things when we really want/need them in this modern country. I'm experimenting with Coral TPU and some open source captcha solvers at the moment, stay tuned. I'm sure ESB is aware of this repo and it took them over 1year to realise whats going on & close the "loophole".

It's maddening - they don't even distinguish between enterprise/corporate accounts. We have dozens of locations with smart meters and no way to pull down the telemetry for them that wouldn't involve running multiple instances of this script and jumping through hoops every time ESBN make a change to their portal.

**Comment by Jesus-M** [comment]
100%

I would totally understand if they provided an API to download the files and set a limit, even a restrictive one like once every 24h or so, but no way at all to automate the collection makes no sense.

ideally they should open the smart metres to be polled locally!

**Comment by SeanCondon** [comment]
I'm seeing the error when trying to rung this script. I guess they've changed something in the Azure B2C config :-(
```
AADB2C90255: The claims exchange specified in technical profile 'SelfAsserted-LocalAccountSignin-Email-MIG' did not complete as expected. You might want to try starting your session over from the beginning.
```

**Comment by badger707** [comment]
Some late night progress - I think we're in the game folks, almost there...
![image](https://github.com/user-attachments/assets/e814f2ea-4bb6-42d9-852a-543c7280e4c9)

**Comment by badger707** [comment]
Update - login script works, I'm able to reach landing/welcome page without triggering captcha. Captcha is still a thing and will triggered if too many attempts or too many open sessions. Going to work now on file download part, the process/steps for it are slightly changed.
![image](https://github.com/user-attachments/assets/609c181e-22b4-4292-82e1-94d11afc1816)

**Comment by piotrasd** [comment]
wow good news, waiting for new version :) thanks for support

**Comment by badger707** [comment]
@dermotw -- how many MPRNs you have/need to collect the data from? All on separate EBS accounts?

**Comment by badger707** [comment]
Script updated & tested, see Known Limitations before running.

**Comment by badger707** [comment]
Weird issues is surfacing up with this -- script was built in Jupyter and moved to python file to share here, so I'm now testing my integration and python file fails, but Jupyter works just fine and file gets downloaded. I need to look at why python fails and jupyter works.

**Comment by Jesus-M** [comment]
I used it today and it worked fine (I had to fine tune the output for my specific use case).

One suggestion tho, all the 'debug prints' you have could be changed to print only if a flag is enable. That way if one nees the HDF file 'clean' can disable the flag instead of removing/commenting all prints.

For example, line 11

`print("##### REQUEST 1 -- GET [https://myaccount.esbnetworks.ie/] ######")`

to be changed to

`print("##### REQUEST 1 -- GET [https://myaccount.esbnetworks.ie/] ######") if  Debug == True`

This way one can control whether having debug messages or not with one single variable at the beginning of the script. Just an idea.

**Comment by badger707** [comment]
@Jesus-M -- thanks for confirming! Re prints -- yeah, I just later realized that I have left them sitting there, sorry, I will update the code with suggested debug twist.

**Comment by badger707** [comment]
Code updated with debug mode to enable/disable print messages.

**Comment by badger707** [comment]
Code updated with improved error handling when script cant login based on server response/content.

**Comment by SalimSwift** [comment]
Thanks @badger707, really appreciate your work. Is the script suppose to download the file? it doesnt in my case and just prints the results into terminal.

**Comment by badger707** [comment]
@SalimSwift -- it receives the file and prints its content to console to indicate that data is there to grab. Based on your needs, you want either to pass this data to another automation for further processing (thats what most of us do) or you can just simply save it to disk, something like this:
`with open("./www/esb_energy_data_raw.json","w",encoding="utf-8") as jsonf:
      jsonf.write(json_out)`

**Comment by SalimSwift** [comment]
That's what I thought after further reading the code :) I have successfully added the downloaded file to my HA instance. Thanks again

**Comment by badger707** [comment]
@SalimSwift -- did you run the code inside of HA?

**Comment by SalimSwift** [comment]
proxy pi running my background stuff, sent the file to HA and ingested data with the file integration

https://www.home-assistant.io/integrations/file > note that it only run a tail on the file, you will need to invert it.

**Comment by ShaneHastings** [comment]
Thanks for your work on this @badger707. I tested the updated version just now and its working great. Now to scale down the freqency the job runs on my side :) 

<img width="530" alt="Screenshot 2025-01-05 at 15 30 34" src="https://github.com/user-attachments/assets/53ec0d5e-a54c-4952-b332-ac8855a78864" />

**Comment by Jesus-M** [comment]
Hi,

this last version is not working for me.

2 days ago I tried it and I got: 

> [FAILED] Unable to get full set of required cookies from [request_3_response.content] -- too many retries (captcha?) or prior sessions was not closed properly. Please wait 6 hours for server to timeout and try again.

I thought it was maybe because I had made too many tests and I was triggering the captcha so I logged in manually to make sure I had a clean one, and I made sure to log off properly.  And waited.

Yesterday his morning 6:45 h same message. I thought it was because it was less than 24 hours since last attempt, so again I logged in manually with all the stuff, and waited.

Today 6:45 h, same message. So it looks something is not working. 

I have enabled the 'debug' and everything seems to be going well until:

```
##### TEST IF SUCCESS ######
[PASS] SUCCESS -- ALL OK [PASS]
[!] Page Title ::  Logging in...
##### REQUEST - SOUP - state & client_info & code ######
[FAILED] Unable to get full set of required cookies from [request_3_response.content] -- too many retries (captcha?) or prior sessions was not closed properly. Please wait 6 hours for server to timeout and try again.
```

Tomorrow will try again.

**Comment by badger707** [comment]
@Jesus-M -- yeah, have seen this before, this is caused by page asking to click "continue" button. It seems that request_#2 receives unexpected response (stuck on "click button" prompt or something like that) and by trying to ignore that with another request_#3 it receives incomplete response, script is expecting to catch 3 cookies but seeing/receiving only 1.

**Comment by badger707** [comment]
@Jesus-M - check `print("[!] Request #2 text :: ", request_2_response.text)`, this should give/print `{"status":"200"}` .

Also, the error you see will be triggered with incorrect credentials, request_2_response.text will show whats going on. Thats from my observations...

**Comment by Jesus-M** [comment]
And today it worked ... 

I will continue making some tests, slow due to the access limitations, but if there is anything new, I will let you know.

Thanks!

**Comment by pravishj** [comment]
@badger707, based on your repo i have made some ui for HA. https://github.com/pravishj/esb-smart-meter-integration
I am new to both HA and open source contribution. Happy for you to use ....
I was struck on the user login captcha.

In the mean while i will test this as well. Thank you for the efforts.

**Comment by Jesus-M** [comment]
@pravishj do you think these sensors will ever have actual info given the usual delay on ESB HDF?

```
sensor.esb_electricity_usage_today
sensor.esb_electricity_usage_last_24_hours
```

Are you planning to add a button or an 'action' to download the full HDF? That could be interesting, if it can be automated in HA.

**Comment by pravishj** [comment]
@Jesus-M, plan is to use the Energy dashboard and I am also looking to track the exported units as well but that will be in phase2.

**Comment by barreeeiroo** [comment]
I was looking for an ESB integration for my Home Assistant and came across this issue. After seeing the reCaptcha in ESB, I decided to follow a different path and instead try to get the data from my electricity provider: Electric Ireland.

I wrote this small HACS integration that does not have any reCaptcha issues: https://github.com/barreeeiroo/Home-Assistant-Electric-Ireland.

Basically, every hour it fetches the data from Electric Ireland (which is fed from ESB), and it provides the same granularity: every 30 minutes. However, it also provides a metric for the cost: depending on the tariff cost may be different, so it will ingest kWh data on 30 minute intervals, and EUR on 60 minute intervals (Electric Ireland API does not allow to get it every 30 minutes).
